### PR TITLE
ETS optimizations: Configurable and automatically adjustable number of locks for set, bag, and duplicate_bag

### PR DESCRIPTION
**Pull request by kjellwinblad**
Before the commits in this pull request, ETS tables of types set, bag and duplicate_bag used a fixed number of locks to protect the hash buckets from concurrent access. The number of locks was 64 by default but could be changed up to 256 by setting a special flag when compiling the Erlang run-time system. Unfortunately, this is not enough to provide good scalability on modern multi-core machines. This pull request introduces new ETS table options that let the user configure the number of locks used per table or configure a table to automatically adjust the number of locks based on the detected concurrency level.

* Commit 1 (f6e1cb0) Make it possible to set the number of locks used for an ETS table of type set, bag, or duplicate_bag by configuring the table with {write_concurrency, N}, where N is the number of locks.
* Commit 2 (cb21fa4) Add the ETS table configuration option {write_concurrency, auto} that causes the table to detect contention in its locks. The number of locks automatically grows and shrinks based on the detected contention.
* Commit 3 (f6e1cb0) Make {decentralized_counters, true} default for all table types when the setting {write_concurrency, auto} and {write_concurrency, N} is active. This makes sense since centralized counters quickly become a bottleneck on big server machines where the new options are most useful. We did not dare to make {decentralized_counters, true} for tables with {write_concurrency, true} before as decentralized counters can make existing applications that frequently call ets:info(T, size) and ets:info(T, memory) perform much worse, but it makes sense to make {decentralized_counters, true} default for the new options.

Benchmark results for the new table options on a machine with 64 hardware threads are available here:

https://winsh.me/bench/ets_config_locks/ets_bench_result_lock_config.html

* Click on the "Fill screen" buttons to make the graphs larger
* One can hide/show lines by clicking on their labels in the legend of the graphs

### CODE REVIEW

File: erts/emulator/beam/erl_db.c [hunk 1]
```diff
@@ -2254,8 +2254,10 @@ BIF_RETTYPE ets_new_2(BIF_ALIST_2)
     Sint keypos;
     int is_named, is_compressed;
     int is_fine_locked, frequent_read;
+    int no_locks;
```
[the remark this answers is not in the document]
Comment (kjellwinblad) [reply erts/emulator/beam/erl_db.c:2258]: I agree. I will change the name to number_of_locks.

File: erts/emulator/beam/erl_db_hash.c [hunk 2]
```diff
@@ -807,25 +808,54 @@ int db_create_hash(Process *p, DbTable *tbl)
...
+    	 * buckets involved in a grow/shrink operation it protected by
+    	 * the same lock:
 	 */
-	ASSERT(erts_atomic_read_nob(&tb->nactive) % DB_HASH_LOCK_CNT == 0);
+	ASSERT((erts_atomic_read_nob(&tb->szm) + 1) % tb->nlocks == 0);
+        ASSERT(tb->nlocks <= erts_atomic_read_nob(&tb->nactive));
+        ASSERT(erts_atomic_read_nob(&tb->nactive) <= tb->nslots);
+        ASSERT(tb->nslots <= (erts_atomic_read_nob(&tb->szm) + 1));
+        ASSERT((tb->nlocks % 2) == 0);
+        ASSERT((erts_atomic_read_nob(&tb->szm) + 1) % 2 == 0);
```
[the remark this answers is not in the document]
Comment (kjellwinblad) [reply erts/emulator/beam/erl_db_hash.c:1051]: Good suggestion. I will fix that.

File: erts/emulator/beam/erl_db.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (kjellwinblad) [reply erts/emulator/beam/erl_db.c]: Fixed 49c2445

File: erts/emulator/beam/erl_db_hash.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (kjellwinblad) [reply erts/emulator/beam/erl_db_hash.c]: Fixed 49c2445

File: erts/emulator/beam/erl_db_hash.c [hunk 3]
```diff
@@ -807,25 +808,54 @@ int db_create_hash(Process *p, DbTable *tbl)
     sys_memset(tb->first_segtab[0], 0, SIZEOF_SEGMENT(FIRST_SEGSZ));
 
     erts_atomic_init_nob(&tb->is_resizing, 0);
+    if (tb->nlocks == -1 || !(tb->common.type & DB_FINE_LOCKED)) {
+        /*
+          The number of locks needs to be set even if fine grained
+          locking is not used as this variable is used when iterating
+          over the table
+        */
+        tb->nlocks = DB_HASH_LOCK_CNT;
```
[the remark this answers is not in the document]
Comment (kjellwinblad) [reply erts/emulator/beam/erl_db_hash.c:1010]: Fixed and found a test case that depended on iteration order of set (which I also fixed) 49c2445

**Comment by kjellwinblad** [comment]
For convenience I have put the compiled version of the doc after this PR here:

https://winsh.me/bench/ets_conf_lock_doc/lib/stdlib-3.16/doc/html/ets.html#new-2

(The changes are in the ets:new/2 function)

**Comment by kjellwinblad** [comment]
The Erlang/OTP team at Ericsson has decided the following about this PR:

* The option for deciding the exact number of locks for a table shall be removed from the public interface
* The {write_concurrency, auto} option should be kept and decentralized_counters should be active by default when the auto option is enabled.

**Review by sverker — APPROVED** [review]

**Pull request merged**